src/binary-encoding.adoc

```diff
@@ -0,0 +1,1600 @@
+
+[[booleans]]
+== Booleans
+
+`0x5E` represents boolean `true`, while `0x5F` represents boolean `false`.
+
+// XXX: Structurally, these examples should be at depth 3. However, all other example sections in the doc
+//      are at heading depth 4, which keeps them out of the TOC. These are also at depth 4 for consistency.
+
+==== Example encoding of `true`
+[source]
+----
+5e
+----
+
+==== Example encoding of `false`
+[source]
+----
+5f
+----
+
+[[numbers]]
+== Numbers
+
+[[integers]]
+=== Integers
+
+Opcodes in the range `0x50` to `0x58` represent an integer. The opcode is followed by a <<fixedint, `FixedInt`>> that
+represents the integer value. The low nibble of the opcode (`0x_0` to `0x_8`) indicates the size of the `FixedInt`.
+Opcode `0x50` represents integer `0`; no more bytes follow.
+
+Integers that require more than 8 bytes are encoded using the variable-length integer opcode `0xF5`,
+followed by a
+<<varuint, VarUInt>> indicating how many bytes of representation data follow.
+
+==== Example encoding of `0`
+[source]
+----
+┌──── Opcode in 50-58 range indicates integer
+│┌─── Low nibble 0 indicates
+││    no more bytes follow.
+50
+----
+
+==== Example encoding of `17`
+[source]
+----
+┌──── Opcode in 50-58 range indicates integer
+│┌─── Low nibble 1 indicates
+││    a single byte follows.
+51 11
+    └── FixedInt 17
+----
+
+==== Example encoding of `-944`
+[source]
+----
+┌──── Opcode in 50-58 range indicates integer
+│┌─── Low nibble 2 indicates
+││    that two bytes follow.
+52 50 fc
+   └─┬─┘
+FixedInt -944
+----
+
+==== Example variable-length encoding of `-944`
+[source]
+----
+┌──── Opcode F5 indicates a variable-length integer, VarUInt length follows
+│   ┌─── VarUInt 2; a 2-byte FixedInt follows
+│   │    that two bytes follow.
+F5 05 50 fc
+      └─┬─┘
+   FixedInt -944
+----
+
+[[floats]]
+=== Floats
+
+Float values are encoded using the IEEE-754 specification, and can be serialized in four sizes:
+
+* 0 bits (0 bytes), representing the value 0e0 and indicated by opcode `0x5A`
+* 16 bits (2 bytes, link:https://en.wikipedia.org/wiki/Half-precision_floating-point_format[half precision]),
+indicated by opcode `0x5B`
+* 32 bits (4 bytes, link:https://en.wikipedia.org/wiki/Single-precision_floating-point_format[single precision]),
+indicated by opcode `0x5C`
+* 64 bits (8 bytes, link:https://en.wikipedia.org/wiki/Double-precision_floating-point_format[double precision]),
+indicated by opcode `0x5D`
+
+Note that in the Ion data model, float values are always 64 bits. However, if a value can be losslessly serialized
+in fewer than 64 bits, applications may choose to do so.
+
+==== Example encoding of `0e0`
+[source]
+----
+┌──── Opcode in range 5A-5D indicates a float
+│┌─── Low nibble A indicates
+││    a 0-length float; 0e0
+5A
+----
+
+==== Example encoding of `3.14`
+[source]
+----
+┌──── Opcode in range 5A-5D indicates a float
+│┌─── Low nibble B indicates a 2-byte float
+││
+5B 42 47
+   └─┬─┘
+half-precision 3.14
```

Comment:
We have to be a little careful here with the examples, if `3.14` (I didn't check) is irrational with respect to base-2, the actual rational value has to be indicated for the smaller formats.

See: https://amazon-ion.github.io/ion-docs/docs/float.html

> In text, binary float is represented using familiar base-10 digits. While this is convenient for human representation, there is no explicit notation for expressing a particular floating point value as binary32 or binary64. Furthermore, many base-10 real numbers are irrational with respect to base-2 and cannot be expressed exactly in either binary floating point encoding (e.g. 1.1e0).
>
> Because of this asymmetry, the rules for Ion text float notation when round-tripping to Ion binary MUST be observed:
>
> * Any text notation that can be exactly represented as binary32 MAY be encoded as either binary32 or binary64 in Ion binary.
> * Any text notation that can only be exactly represented as binary64 MUST be encoded as binary64 in Ion binary.
> * Any text notation that has no exact representation (i.e. irrational in base-2 or more precision than the binary64 mantissa), MUST be encoded as binary64. This is to ensure that irrational numbers or truncated values are represented in the highest fidelity of the float data type.

Reply:
I had confirmed that `3.14` could be roundtripped losslessly through `f32` and `f64` using the [Float Toy](https://evanw.github.io/float-toy/) website.